include/ygm/io/parquet_parser.hpp

```diff
@@ -54,60 +78,86 @@ std::ostream& operator<<(std::ostream& os, const parquet_data_type& t) {
 // Only supports the plain encoding.
 // Do not support nested or hierarchical columns.
 class parquet_parser {
- public:
+ private:
   using self_type = parquet_parser;
-  // 0: a column type, 1: column name
-  using file_schema_container =
-      std::vector<std::tuple<detail::parquet_data_type, std::string>>;
-  using parquet_stream_reader = parquet::StreamReader;
 
-  parquet_parser(ygm::comm& _comm) : m_comm(_comm), pthis(this) {
-    pthis.check(m_comm);
-  }
+ public:
+  // List of C++ types supported by this parser
+  // Note: there is no uint types in Parquet
+  using parquet_type_variant =
+      std::variant<std::monostate, bool, int32_t, int64_t, float, double,
+                   std::string>;
+
+  struct column_schema_type {
+    detail::parquet_data_type type;
+    std::string               name;
+    // If true, this parser can not handle this column data.
+    // This can happen due to unsupported data type or non-flat column structure.
+    bool unsupported{false};
+  };
 
-  parquet_parser(ygm::comm& _comm, const std::vector<std::string>& stringpaths,
-                 bool recursive = false)
+  parquet_parser(ygm::comm &_comm, const std::vector<std::string> &stringpaths,
+                 const bool recursive = false)
       : m_comm(_comm), pthis(this) {
     pthis.check(m_comm);
-    check_paths(stringpaths, recursive);
-    read_file_schema();
-    m_comm.barrier();
+    init(stringpaths, recursive);
   }
 
   ~parquet_parser() { m_comm.barrier(); }
 
-  // Returns a list of column schema information
-  const file_schema_container& schema() { return m_schema; }
+  // Returns a list of column schema (simpler version).
+  // The order of the schema is the same as the order of Parquet column
+  // indices (ascending order).
+  // This function assumes that all files have the same schema.
+  const std::vector<column_schema_type> &get_schema() { return m_col_schema; }
 
-  const std::string& schema_to_string() { return m_schema_string; }
+  // Return full Parquet file schema directly returned by the Parqet reader as a
+  // string. This function assumes that all files have the same schema.
+  const std::string &schema_to_string() { return m_schema_string; }
 
+  /// Read all columns and call the function for each row.
+  /// fn is called with a std::vector<parquet_type_variant>.
+  /// The values of unsupported columns are set to std::monostate.
   template <typename Function>
+    requires std::invocable<Function, const std::vector<parquet_type_variant> &>
   void for_all(Function fn) {
-    read_files(fn);
+    read_parquet_files(fn);
+  }
+
+  /// for_all(), read only the specified columns.
+  /// The values of the columns that do not exist are set to std::monostate.
+  template <typename Function>
+    requires std::invocable<Function, const std::vector<parquet_type_variant> &>
+  void for_all(const std::vector<std::string> &columns, Function fn) {
+    read_parquet_files(fn, columns);
   }
 
```

Comment:
Should we also support a `for_all` that allows user lambdas to call columns by name instead of forcing positional arguments?

This would be separate from the support you have that prevents reading data that is unneeded.

Reply:
That's a good idea. I'll implement it later if we can wait. For now, I want to keep this API as simple as possible, as we are actively changing this reader class drastically.